posix/iov.c

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <openenclave/bits/safecrt.h>
+#include <openenclave/corelibc/limits.h>
+#include <openenclave/corelibc/stdio.h>
+#include <openenclave/corelibc/stdlib.h>
+#include <openenclave/corelibc/string.h>
+#include <openenclave/corelibc/sys/uio.h>
+#include <openenclave/internal/posix/iov.h>
+
+/* Get the deflated size of the IO vector. */
+size_t oe_iov_compute_size(const struct oe_iovec* iov, size_t iov_count)
+{
+    size_t size = 0;
+
+    for (size_t i = 0; i < iov_count; i++)
+        size += iov[i].iov_len;
+
+    return size;
+}
+
+int oe_iov_deflate(
+    const struct oe_iovec* iov,
+    size_t iov_count,
+    void** buf_out,
+    size_t* buf_size_out)
+{
+    int ret = -1;
+    void* buf = NULL;
+    size_t buf_size = 0;
+
+    if (buf_out)
+        *buf_out = NULL;
+
+    if (buf_size_out)
+        *buf_size_out = 0;
+
+    if (!iov || !buf_out || !buf_size_out)
```

Comment:
We should just reject `iovcnt == 0` here as well for clarity since it's not valid for `writev` as it allows short circuiting of a series of no-ops and elides a zero size alloc.

Reply:
From the manpage for writev:

```
       EINVAL The  vector  count  iovcnt is less than zero or greater than the
              permitted maximum.
```

As a general rule, zero-sized writes and reads are legal. This returns 0 and does not raise an errno:

```
writev(STDERR_FILENO, NULL, 0);
```

Reply:
It's interesting that it doesn't align with the [POSIX](http://pubs.opengroup.org/onlinepubs/9699919799/functions/writev.html) spec:

> The writev() function shall be equivalent to write(), except as described below. The writev() function shall gather output data from the iovcnt buffers specified by the members of the iov array: iov[0], iov[1], ..., iov[iovcnt-1]. The **iovcnt argument is valid if greater than 0** and less than or equal to {IOV_MAX}, as defined in <limits.h>.

Similarly for [readv](http://pubs.opengroup.org/onlinepubs/9699919799/functions/readv.html).

The comment was made because of the inconsistency noted in your example. If we want to support the manpage semantics, that's fine as well so long as we test for that consistency.

Reply:
thing is that Linux is "mostly POSIX-compliant" so you will see these inconsistencies. I'm in favor of implementing it to the manpage since the code is running on Linux. I think the name `liboesyscall` more actually reflects what this layer is.

Reply:
For my own understanding, what is the manpage for writev that we are using as a reference for the behavior of these functions?

The following is the output I get on my Ubuntu VM:
```bash
$ man writev | grep EINVAL
       EINVAL The sum of the iov_len values in the iov array would overflow an ssize_t.
       EINVAL The iovcnt argument was less than or equal to 0, or greater than {IOV_MAX}.
```
which does not match what @mikbras is referring to for iovcnt==0:
> EINVAL The  vector  count  iovcnt is less than zero or greater than the
              permitted maximum.

Reply:
That's right: I was referring to "man 2 writev" (same as what you have). We comply with the statement you pasted above.

Reply:
I think we have to tolerate iovcnt == 0 since Linux does.

Reply:
I would advise against rejecting a zero-valued **iovcnt** since it is permitted by the Linux kernel, and Linux is our target environment.

Reply:
@mikbras Thanks for the clarification! On my Manjaro desktop, _man writev_ takes me to the Posix Programmer's Manual whereas _man 2 writev_ takes me to the  Linux Programmer's Manual :laughing:

Reply:
@gupta-ak What do you mean by "since the code is running in Linux"?   OE is supposed to be host OS agnostic.

Reply:
I think it would be a mistake to force non-POSIX Linux-only behavior on non-Linux hosts.   I think it's fine to force non-POSIX Linux-only behavior on Linux hosts.  But OE needs to be written to be agile.   It's fine to have an ifdef or something to control the behavior, but don't put Linux-specific code in a global non-ifdef'ed location.

Reply:
I agree that we want to be agile. It is more  agile to tolerate a zero-valued iovcnt in **writev** and **readv** rather than to reject it outright. We could reject this outright, but it might break some applications. Accepting it is not likely to break any platform. We could ifdef on the enclave side.

Reply:
@dthaler OE is host agnostic, but this code (and everything in `posix/`) is still in the enclave and all the enclave code assumes it's running on Linux.

For example, let's say you're an application developer that calls the libc/POSIX function called `open`. Our libc library (MUSL) will implement this function by doing a `SYS_open` Linux system call. Since that will not obviously work in the enclave, Mike's code will intercept this system call and implement it through an OCALL. The actual OCALL  implementation itself is where the host can be Windows or Linux.

I think there's a lot of confusion since the Linux system calls have pretty similar names to the POSIX functions. Mike's PR is mainly on implementing the **Linux syscall ABI**, so `oe_open` implements `SYS_open` NOT the `open` POSIX/C function. The POSIX/C function is implemented by MUSL so we don't need to reimplement it. This way we get a bunch of functions that "just work" in the enclave like `htons`.

So it looks like this:
`user app -> MUSL POSIX + C library -> Linux syscall ABI -> OCALL on Linux/Windows/other OS`.

Most of the POSIX compliance should be handled in the C library itself, so the system call ABI doesn't need to be strictly POSIX compliant.

Reply:
Nice write up @gupta-ak.

Yes, Linux is **mostly** POSIX compliant; but not POSIX certified.

Some clarification to your write up.

> user app -> MUSL POSIX + C library -> Linux syscall ABI -> OCALL on Linux/Windows/other OS.

This is accurate and a great description. However, unfortunately, it is not complete.
We also have another way to call the implementations directly via corelibc which completely bypasses the syscall dispatcher mechanism.
https://github.com/microsoft/openenclave/blob/2316cc983824d6d3bb39ba63691015be8d5fef37/include/openenclave/corelibc/fcntl.h#L131-L134

So we also have this other path of calling the oe_* functions in this PR:

`??? -> corelibc -> oe_* function`

As you can this, this implies moving implementation of Linux syscall ABI directly into corelibc, bypassing the syscall dispatcher framework, making it unclear whether oe_open should confirm to POSIX or GLIBC or Linux syscall semantics.

Although POSIX, GLIBC and Linux syscall ABI are very similar, they have subtle and not so subtle differences.

Consider an interesting albeit complicated example of the Linux syscall ABI function  **signal**.
Unlike usual GLIBC functions that are thin wrappers over the corresponding Linux Syscall function, the GLIBC signal function does not call the **Sys_signal** syscall; instead it dispatches to **Sys_sigaction** syscall.

```bash
$man 2 signal | grep sigaction
By  default,  in glibc 2 and later, the signal() wrapper function does not invoke the kernel system call.  Instead, it calls sigaction(2) using flags that supply BSD semantics.  This default behavior is provided as long as a
         suitable feature test macro is defined: _BSD_SOURCE on glibc 2.19 and earlier or _DEFAULT_SOURCE in glibc 2.19 and later.  (By default, these macros are defined; see feature_test_macros(7) for details.)  If  such  a  feature
         test macro is not defined, then signal() provides System V semantics.
```
Here is MUSL's implementation of signal which mirrors GLIBC and dispatches to sigaction.
https://github.com/microsoft/openenclave/blob/2316cc983824d6d3bb39ba63691015be8d5fef37/3rdparty/musl/musl/src/signal/signal.c#L4-L10

And here is the implementation of signal in corelibc which deviates from GLIBC and dispatches to signal syscall:
https://github.com/microsoft/openenclave/blob/2316cc983824d6d3bb39ba63691015be8d5fef37/include/openenclave/corelibc/signal.h#L174-L177

At this point it becomes quite difficult to reason about the behavior of enclaves, because a call to 'signal' will have different semantics depending on whether MUSL implementation is being used or corelibc behavior is being used. And these differences are lurking around in the code.

The subtle differences in API behavior may not matter much to the average programmer.
However, if you are implementing a real time system and want to analyze the fault tolerance of the system under error conditions, then strict conformance to a published standard matters.
Once standards exist and are confirmed to, then automated static and dynamic analysis tools can be used to rigorously test and analyze the behavior of a system.

Personally, I don't think that it matters right now which standard subset we confirm to (POSIX or GLIBC or Linux Syscall ABI). Most likely it should be easier to change from one standard to another. However, we *do* need to confirm to a (subset of a) standard and not have silent behavioral deviations. 

Note: I am not sure whether the operating system layer for Trust Zone Enclaves confirm to POSIX or to some other standard.

I agree that this PR implements the Linux Syscall ABI and therefore liboeposix is a misnomer.

Reply:
Thank you Anand for your comments. We will track the naming as an issue. Hopefully we can find a better name that will not confuse people.